### DISCUSSION TITLE
Pattern Matching on nullable structs are weird

### DISCUSSION THREAD

**Discussion by ziaulhasanhamim** [dotnet/csharplang#6577] [Language Ideas] [answered]
```C#
int? count = ...;
var msg = count switch
{
  null => "count is null",
  var val => DoStuff(val)
};
string DoStuff(int count)
{
   // code
};
```
This code will give a compilation error. Because DoStuff takes `int` as a parameter, but the type of `val` will be `int?`. But isn't it more appropriate to `val` be a 'int' instead of 'int?' because I already checked for the null condition? Also because, `Nullable<T>` isn't treated as a regular struct in pattern matching

**Comment by kyoyama-kazusa** [comment, marked as answer]
`var` pattern always defines a new variable that is exactly of same type as the expression. Because `count` is `int?`, `val` is an `int?` variable.

Although you has already checked the nullability, in nullable struct types, `T?` means `Nullable<T>`, which is a different type with `T`. Therefore you should at least emit a real not-null pattern instead of above given node snippet, for example:

```csharp
int? count = ...;
var msg = count switch
{
    { } val => DoStuff(val),
    _ => "count is null"
};

string DoStuff(int count)
{
    // ...
}
```

[Sharplab](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgyABAJgEYyzYAXAflwGMB7ZGR3AXlwxsAd1xNmACgCUAbjIA3CACdcYAM4Bzfuy49ca4VEZsAFmQDeZXNdzncAX1yLU/AHy4AIhwDKjZADN/CWcpABorGwB9N1wAIk5uXig1QTRUWLJ7OXJSChoABk8fP0CJJh1EqQsI6woAdlw8/IA6AFEwAAdGAE9s+yA=)

We should adjust the null checking, using `{ } variable` instead of `null`.

Of course, you can also use the original way to check:

```csharp
var msg = count switch
{
    null => ...,
    { } val => ...
}
```

[Sharplab](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgyABAJgEYyzYAXAflwGMB7ZGR3AXlwxsAd1xNmACgCUAbjIA3CACdcYAM4Bzfuy49ca4VEZsAFmQDeZXNcFpU/AHy4ARJ268oa2+mcAaKzbmuAC+uIr2fE4AIhwAyozIAGaJEuFSZMFy5KQUNAAMuDHxSSlMOu7ppJakNrgUAOx1+QB0AKJgAA6MAJ5ZwUA=)

**Comment by rbwhitaker** [comment]
I'd suggest this instead, which seems to have the behavior you're looking for and is quite readable/understandable to my eyes.

```cs
int? count = ...;
var msg = count switch
{
    null  => "count is null",
    int n => DoStuff(n)
};
```
